admin/reverseGpx.php

```diff
@@ -43,21 +45,18 @@
 $editfile = $_FILES[$name]['tmp_name'];
 libxml_use_internal_errors(true);
 $adminpg = "../admin/admintools.php";
-$dom = new DOMDocument();
-$dom->formatOutput = true;
-if (!$dom->load($editfile)) {
-    displayGpxUserAlert($gpxfile);
-    header("Location: {$adminpg}");
+validateGpxFile($editfile, $gpxfile);
+if (isset($_SESSION['usr_alert'])) {
+    header("Location: " . $adminpg);
```

Comment:
Why do we need the variable $adminpg when it always contains the same value?

Reply:
OK. It took a long time for me to converge on the following long response. I hope it doesn't come off as nitpicking. I think most if not all of my concerns can be addressed be agreeing on the high-level structure for the code before doing all the implementation.

### We should code constants as constants
If `$adminpg` is a constant, then I think we should code it as a constant. That way the person reading the code can understand the intent. So instead of:
`$adminpg = '../admin/admintools.php';`
do something like:
`const ADMINPG = '../admin/admintools.php';`
or put the whole string into the constant:
`const ADMINPG = 'Location: ../admin/admintools.php';`


### But I think there is a still better way to go
If the goal is to reduce the amount of duplicated code and improve maintainability, couldn't we move the whole if block into validateGpxFile()? We would just need to add the redirect location as a parameter, like this:
`validateGpxFile($editfile, $gpxfile, ADMINPG);`
But I don't think we even need to create our own constant. PHP has one for us:
`validateGpxFile($editfile, $gpxfile, __FILE__);`


### We should improve the file name tests
We do some generic file name tests at the beginning of validateUpload(). We check that the file name is not blank and that it contains a period.
1. Given that the user is picking a file from a list generated by the browser, would the existing tests ever fail?
1. If we do think we need some tests like this, I think we should:
   1. Enhance them, as currently we would accept a file name consisting of a single period.
   1. Consider doing the filename tests in js, as this feels like part of validating the user's input on the form. Doing it that way would give the user immediate feedback and save an exchange with the server.


### Proposed alternative structure for what used to be validateUpload()
Given the above, I think we should further restructure the code into the following reusable parts:
- genericFilenameTests() - Check that the name of the uploaded file conforms to our requirements.
  - I think existing Issue: #8 relates to this.
  - See **We should improve the file name tests** above
- gpxFileTests() - Two parts: Name tests and Contents tests
  - GPX file name tests: check that file name ends in .gpx or .GPX. Along similar lines to the notes above, consider whether to:
    - Do the test at all. Would the site fail on a gpx file whose name does not end with .gpx?
    - Do the test in js, as this feels like part of validating the user's input on the form.
  - GPX file contents tests: Load XML and validate against the GPX schema
    - Could/Should we do these tests in js?
- moveUploadedFile() - move uploaded files that are to be permanently stored on the site.
  - I think we need to enhance this code to protect against race conditions that could result in multiple files with the same name being created. I've written a separate issue for this: #12.

Reply:
Impressive thought put in on this - well done!
There are several 'sections' to your comment, I will take them as listed:
1. **Constants as constants**: understood, makes sense.
2. **Better way to go**: can you provide a little more detail - are you saying we create a new function, validateGpxFile(..), or are you using that name as a generic for the displayGpxUserAlert? If new, would this then execute both xml and gpx schema validation (and maybe later <ele> check)? I thought we wanted to maintain these checks as 'atomic', but maybe they always all get done together? I'm assuming we still want to only return the first error encountered. If new, where would the new function reside? Currently, we have most of our upload functions in buildFunctions.php. Remember that *__FILE__* provides the filename of the *currently* executing script - i.e. reverseGpx.php or buildFunctions.php. We can't redirect to those - they are not pages. Also, for saveTab1, we would have to add the hike no in the redirect parameter, as the redirect to editDB requires the hike no in the query string. That parameter would now have to be added to the argument list of validateUpload, which it otherwise has no need for.  It seems to increase complexity as you add arguments.
3. **File name tests**: **(1)** I pick the defective file for testing from the browser's list. A user could certainly fat finger his choice, or received a bad file from someone else. In addition to mild name checks, we also test that the server upload was successful. It seems we do need some checks. **(2)** If we do need checks - yes, we should enhance them. That really sounds like a separate issue: would you like to report that as a newbie? Js tests of the filename make good sense. I don't think this is covered in any existing issue.
4. **Alternative Structure**: **(1)** Generic filename tests: see Item 3 above. I'm not convinced #8 addresses this adequately. **(2)** Gpx file tests: for names, yes, we covered this in previous remark above. We already test for gpx or GPX extension. I don't doubt it would fail if it does not have the correct extension - but it really should have it. I treat that as another rather mild security check - it would be suspicious if it doesn't have the right extension. It might, for instance, point out a fat finger mis-select. Content checking is a very difficult thing for js. First of all, the file isn't alredy uploaded on the page via selection, but by the server when the form is submitted. Secondly, there don't seem to be good ways to do these checks in js without importing 3rd party packages, and I don't know that they cover our needs. In fact, the one I looked at was an npm module - which we aren't currently incorporating as they are intended mostly for server-side code. Most of the methods I saw online (limited) use ajax to bring the file in first. More code to do something that is simple now. Name checking in js makes sense, content checking doesn't. Php does this with a single command, no extra packages required. Why not use this existing capability. **(3)** moveUploadedFile - addressed in #12.